In [176]:
# testing pcap upload function
from nfstream import NFStreamer
import os

def load_and_process_pcap(pcap_file_path):
    # Sprawdzenie, czy plik istnieje
    if not os.path.exists(pcap_file_path):
        print(f"Plik {pcap_file_path} nie istnieje.")
        return

    try:
        streamer = NFStreamer(source=pcap_file_path, statistical_analysis=True)
        for flow in streamer:
            print(flow.bidirectional_mean_ps)  
    except Exception as e:
        print(f"Wystąpił problem podczas przetwarzania pliku {pcap_file_path}: {e}")

pcap_file_path = '../data/raw/testing_small.pcapng'

load_and_process_pcap(pcap_file_path)


96.0
71.0
90.0
91.0
179.0
333.0552995391705
107.18705035971226
116.0
470.0
54.0
71.0
91.0
106.25259515570929
415.14782608695657


In [175]:
# testing pcap upload function
from nfstream import NFStreamer
import os

def load_and_process_pcap(pcap_file_path):
    # Sprawdzenie, czy plik istnieje
    if not os.path.exists(pcap_file_path):
        print(f"Plik {pcap_file_path} nie istnieje.")
        return

    try:
        streamer = NFStreamer(source=pcap_file_path, statistical_analysis=True)
        for flow in streamer:
            print(flow.bidirectional_mean_ps)  
    except Exception as e:
        print(f"Wystąpił problem podczas przetwarzania pliku {pcap_file_path}: {e}")

perturbed_file_path = '../data/interim/testing_small_perturbed.pcapng'

load_and_process_pcap(pcap_file_path)


96.0
71.0
470.0
54.0
90.0
91.0
107.18705035971226
179.0
116.0
333.0552995391705
71.0
91.0
106.25259515570929
415.14782608695657


In [205]:
from scapy.all import PcapReader, IP, TCP, UDP

def process_packets(file_path):
    with PcapReader(file_path) as pcap_reader:
        for packet_number, packet in enumerate(pcap_reader):
            timestamp_ms = packet.time * 1000
            srcip = dstip = srcport = dstport = proto = None

            if IP in packet:
                srcip = packet[IP].src
                dstip = packet[IP].dst
                proto = packet[IP].proto  # Numer protokołu

                if TCP in packet:
                    srcport = packet[TCP].sport
                    dstport = packet[TCP].dport
                elif UDP in packet:
                    srcport = packet[UDP].sport
                    dstport = packet[UDP].dport

            # Wyświetlenie informacji o pakiecie
            print(f"Packet Number: {packet_number + 1}")  # +1, jeśli chcemy numerację od 1
            print(f"Timestamp (ms): {timestamp_ms}")
            print(f"Source IP: {srcip}, Destination IP: {dstip}")
            print(f"Source Port: {srcport}, Destination Port: {dstport}")
            print(f"Protocol Number: {proto}\n")

# Użycie funkcji
process_packets(pcap_file_path)


Packet Number: 1
Timestamp (ms): 1595284457571.949000
Source IP: 192.168.56.1, Destination IP: 192.168.56.101
Source Port: 59281, Destination Port: 3389
Protocol Number: 6

Packet Number: 2
Timestamp (ms): 1595284457572.13000
Source IP: 192.168.56.101, Destination IP: 192.168.56.1
Source Port: 3389, Destination Port: 59281
Protocol Number: 6

Packet Number: 3
Timestamp (ms): 1595284457572.317000
Source IP: 192.168.56.1, Destination IP: 192.168.56.101
Source Port: 59281, Destination Port: 3389
Protocol Number: 6

Packet Number: 4
Timestamp (ms): 1595284457573.703000
Source IP: 192.168.56.1, Destination IP: 192.168.56.101
Source Port: 59281, Destination Port: 3389
Protocol Number: 6

Packet Number: 5
Timestamp (ms): 1595284457624.09000
Source IP: 192.168.56.101, Destination IP: 192.168.56.1
Source Port: 3389, Destination Port: 59281
Protocol Number: 6

Packet Number: 6
Timestamp (ms): 1595284457648.739000
Source IP: 192.168.56.101, Destination IP: 192.168.56.1
Source Port: 3389, Destinat

In [208]:
type(packet)

__main__.TruncatedPacket

In [ ]:
# scapy:

#timestamp ms
#srcip, dstip, srcport, dstport, proto(number)

In [46]:
#testing packet
cap = pyshark.FileCapture(file_path, display_filter='ip and (tcp or udp)') # we will operate only on ip packets
packet = cap[1]
packet

<TCP Packet>

In [47]:
#testing nfflow
streamer = NFStreamer(source=pcap_file_path)
flows = []
for flow in streamer:
    flows.append(flow)
    
flow = flows[3]
flow

In [48]:
type(packet)

pyshark.packet.packet.Packet

In [195]:
import nfstream
import scapy

def generate_pseudo_hash(entity):
    if isinstance(entity, nfstream.flow.NFlow):
       elements = [hash(entity.src_ip), hash(entity.dst_ip), int(entity.src_port), int(entity.dst_port), int(entity.protocol)] 
       #print(entity.src_ip)
       #print(entity.dst_ip)
       #print(entity.src_port)
       #print(entity.dst_port)
       #print(elements)
    
    elif isinstance(entity, pyshark.packet.packet.Packet): 
        elements = [hash(entity.ip.src), hash(entity.ip.dst), int(entity[entity.transport_layer].srcport), int(entity[entity.transport_layer].dstport), int(entity.ip.proto)]
        #print(entity.ip.src)
        #print(entity.ip.dst)
        #print(entity[entity.transport_layer].srcport)
        #print(entity[entity.transport_layer].dstport)
        #print(elements)
    
    elif isinstance(entity, scapy.layers.l2.Ether):
        if entity.haslayer(TCP):
            elements = [hash(entity[IP].src), hash(entity[IP].dst), int(entity[TCP].sport), int(entity[TCP].dport), int(entity[IP].proto)] 
        elif entity.haslayer(UDP):
            elements = [hash(entity[IP].src), hash(entity[IP].dst), int(entity[UDP].sport), int(entity[UDP].dport),  int(entity[IP].proto)] 
    else:
        print("ERROR: not a flow nor packet")

    return sum(elements)

In [179]:
type(packet)

scapy.layers.l2.Ether

In [50]:
packet.ip.proto

'6'

In [51]:
packet[packet.transport_layer].dstport

'59281'

In [207]:
type(packet)

__main__.TruncatedPacket

In [196]:
generate_pseudo_hash(packet)

ERROR: not a flow nor packet


UnboundLocalError: cannot access local variable 'elements' where it is not associated with a value

In [57]:
generate_pseudo_hash(flow)

16907088091671947950

In [54]:
# pseudo-hash function testing

for flow in flows:
    x = 0
    for packet in cap:
        if (generate_pseudo_hash(flow) == generate_pseudo_hash(packet)):
            x += 1
            print(f'for {flow} found {x} packets')

for NFlow(id=0,
      expiration_id=0,
      src_ip=192.168.56.101,
      src_mac=08:00:27:e6:e5:59,
      src_oui=08:00:27,
      src_port=5353,
      dst_ip=224.0.0.251,
      dst_mac=01:00:5e:00:00:fb,
      dst_oui=01:00:5e,
      dst_port=5353,
      protocol=17,
      ip_version=4,
      vlan_id=0,
      tunnel_id=0,
      bidirectional_first_seen_ms=1595284458733,
      bidirectional_last_seen_ms=1595284458765,
      bidirectional_duration_ms=32,
      bidirectional_packets=4,
      bidirectional_bytes=384,
      src2dst_first_seen_ms=1595284458733,
      src2dst_last_seen_ms=1595284458765,
      src2dst_duration_ms=32,
      src2dst_packets=4,
      src2dst_bytes=384,
      dst2src_first_seen_ms=0,
      dst2src_last_seen_ms=0,
      dst2src_duration_ms=0,
      dst2src_packets=0,
      dst2src_bytes=0,
      application_name=MDNS,
      application_category_name=Network,
      application_is_guessed=0,
      application_confidence=6,
      requested_server_name=msedgewin10.loc

In [ ]:
# czyli powyzsze sie zgadza, wiec teraz: 1) stworzyc truncated pakiety obejmujace hash[x], 2) pogrupowac trucated pakiety po flow i dodac flow id, 3) zweryfikowac czy sie zgadza

In [245]:
# truncated packet class

class TruncatedPacket:
    def __init__(self, packet_id, timestamp, size, pseudo_hash, flow_id, direction, src_ip):
        self.packet_id = packet_id
        self.timestamp = timestamp
        self.size = size
        self.pseudo_hash = pseudo_hash
        self.flow_id = flow_id
        self.direction = direction
        self.src_ip = src_ip

    def __repr__(self):
        return (f"TruncatedPacket(packet_id={self.packet_id}, "
                f"timestamp={self.timestamp}, "
                f"size={self.size}, "
                f"pseudo_hash='{self.pseudo_hash}', "
                f"flow_id={self.flow_id}, "
                f"src_ip={self.src_ip}, "
                f"direction='{self.direction}')")

In [246]:
# SCAPY TRUNCATED PACKETS GENERATION

def create_truncated_packets_from_pcap(file_path):
    truncated_packets = []
    cap = rdpcap(file_path)
    
    for packet_number, scapy_packet in enumerate(cap, start=1):
        if (scapy_packet.haslayer(TCP) or scapy_packet.haslayer(UDP)) and IP in scapy_packet:
            pseudo_hash = generate_pseudo_hash(scapy_packet)
            if pseudo_hash:  # we make sure that this was calculated successfully
                truncated_packet = TruncatedPacket(
                    packet_id=packet_number * 10, # according to schema
                    timestamp=scapy_packet.time, # timestamp will need to have margin of error
                    size=len(scapy_packet),
                    pseudo_hash=pseudo_hash,
                    flow_id=None,  # later
                    direction=0,  # 0 - not yet analyzed, 1 - fwd, 2 - bwd
                    src_ip=scapy_packet[IP].src
                )
                truncated_packets.append(truncated_packet)
    
    return truncated_packets

file_path = '../data/raw/testing_small.pcapng'
truncated_packets = create_truncated_packets_from_pcap(file_path)

# Przykładowe wyświetlenie pierwszych 5 pakietów
for packet in truncated_packets[:10]:
    print(packet)

TruncatedPacket(packet_id=10, timestamp=1595284457.571949, size=66, pseudo_hash='16907088091671947946', flow_id=None, src_ip=192.168.56.1, direction='0')
TruncatedPacket(packet_id=20, timestamp=1595284457.57213, size=66, pseudo_hash='16907088091671947946', flow_id=None, src_ip=192.168.56.101, direction='0')
TruncatedPacket(packet_id=30, timestamp=1595284457.572317, size=60, pseudo_hash='16907088091671947946', flow_id=None, src_ip=192.168.56.1, direction='0')
TruncatedPacket(packet_id=40, timestamp=1595284457.573703, size=98, pseudo_hash='16907088091671947946', flow_id=None, src_ip=192.168.56.1, direction='0')
TruncatedPacket(packet_id=50, timestamp=1595284457.62409, size=54, pseudo_hash='16907088091671947946', flow_id=None, src_ip=192.168.56.101, direction='0')
TruncatedPacket(packet_id=60, timestamp=1595284457.648739, size=73, pseudo_hash='16907088091671947946', flow_id=None, src_ip=192.168.56.101, direction='0')
TruncatedPacket(packet_id=70, timestamp=1595284457.689521, size=60, pseu

In [248]:
def assign_flow_ids_to_packets(truncated_packets, streamer):
    for flow in streamer:
        for packet in truncated_packets:
            packet_timestamp = float(packet.timestamp) * 1000
            flow_firstseen = float(flow.bidirectional_first_seen_ms)
            flow_lastseen = float(flow.bidirectional_last_seen_ms)
            
            
            if flow_firstseen - 1.0 <= packet_timestamp <= flow_lastseen + 1.0 and packet.pseudo_hash == generate_pseudo_hash(flow):
                packet.flow_id = flow.id
                if packet.src_ip == flow.src_ip:
                    packet.direction = 1
                else:
                    packet.direction = 2



In [249]:
# Użycie funkcji
streamer = NFStreamer(source=pcap_file_path)
assign_flow_ids_to_packets(truncated_packets, streamer)
for packet in truncated_packets:
    print(packet)


TruncatedPacket(packet_id=10, timestamp=1595284457.571949, size=66, pseudo_hash='16907088091671947946', flow_id=5, src_ip=192.168.56.1, direction='1')
TruncatedPacket(packet_id=20, timestamp=1595284457.57213, size=66, pseudo_hash='16907088091671947946', flow_id=5, src_ip=192.168.56.101, direction='2')
TruncatedPacket(packet_id=30, timestamp=1595284457.572317, size=60, pseudo_hash='16907088091671947946', flow_id=5, src_ip=192.168.56.1, direction='1')
TruncatedPacket(packet_id=40, timestamp=1595284457.573703, size=98, pseudo_hash='16907088091671947946', flow_id=5, src_ip=192.168.56.1, direction='1')
TruncatedPacket(packet_id=50, timestamp=1595284457.62409, size=54, pseudo_hash='16907088091671947946', flow_id=5, src_ip=192.168.56.101, direction='2')
TruncatedPacket(packet_id=60, timestamp=1595284457.648739, size=73, pseudo_hash='16907088091671947946', flow_id=5, src_ip=192.168.56.101, direction='2')
TruncatedPacket(packet_id=70, timestamp=1595284457.689521, size=60, pseudo_hash='169070880

In [250]:
def count_packets_by_flow_id(truncated_packets):
    counts = {}
    for packet in truncated_packets:
        # Przypisanie do flow_id wartości None, jeśli flow_id nie jest ustawione
        flow_id = packet.flow_id if packet.flow_id is not None else 'None'
        if flow_id in counts:
            counts[flow_id] += 1
        else:
            counts[flow_id] = 1
    return counts

# Załóżmy, że truncated_packets to twoja lista obiektów TruncatedPacket
counts = count_packets_by_flow_id(truncated_packets)

# Przykład wykorzystania i wydrukowania wyników
for flow_id, count in counts.items():
    print(f"Flow ID: {flow_id}, Count: {count}")


Flow ID: 5, Count: 278
Flow ID: 1, Count: 2
Flow ID: 0, Count: 4
Flow ID: 4, Count: 1
Flow ID: 3, Count: 1
Flow ID: 10, Count: 6
Flow ID: 12, Count: 217
Flow ID: 11, Count: 289
Flow ID: 13, Count: 230


NFS:
- NFlow(id=0): bidirectional_packets=2
- NFlow(id=1): bidirectional_packets=278
- NFlow(id=2): bidirectional_packets=4
- NFlow(id=3): bidirectional_packets=230
- NFlow(id=4): bidirectional_packets=1
- NFlow(id=5): bidirectional_packets=5
- NFlow(id=6): bidirectional_packets=4
- NFlow(id=7): bidirectional_packets=1
- NFlow(id=8): bidirectional_packets=5
- NFlow(id=9): bidirectional_packets=1
- NFlow(id=10): bidirectional_packets=1
- NFlow(id=11): bidirectional_packets=289
- NFlow(id=12): bidirectional_packets=6
- NFlow(id=13): bidirectional_packets=217

In [191]:
import pyshark

import nest_asyncio

# Zastosuj nest_asyncio
nest_asyncio.apply()

def print_packet_sizes(pcap_file):
    # Otwórz plik pcap
    cap = pyshark.FileCapture(pcap_file)

    # Iteruj przez pakiety i wyświetl ich wielkość
    for packet in cap:
        try:
            # Wyświetl długość pakietu
            print(f"Wielkość pakietu: {packet.length} bajtów")
        except AttributeError:
            # W przypadku braku atrybutu length, kontynuuj
            continue
        

modified_pcap_path = "../data/interim/testing_small_perturbed.pcapng"
print_packet_sizes(modified_pcap_path)

Wielkość pakietu: 66 bajtów
Wielkość pakietu: 66 bajtów
Wielkość pakietu: 60 bajtów
Wielkość pakietu: 98 bajtów
Wielkość pakietu: 54 bajtów
Wielkość pakietu: 73 bajtów
Wielkość pakietu: 60 bajtów
Wielkość pakietu: 233 bajtów
Wielkość pakietu: 1257 bajtów
Wielkość pakietu: 1257 bajtów
Wielkość pakietu: 1257 bajtów
Wielkość pakietu: 212 bajtów
Wielkość pakietu: 105 bajtów
Wielkość pakietu: 60 bajtów
Wielkość pakietu: 140 bajtów
Wielkość pakietu: 304 bajtów
Wielkość pakietu: 668 bajtów
Wielkość pakietu: 668 bajtów
Wielkość pakietu: 142 bajtów
Wielkość pakietu: 167 bajtów
Wielkość pakietu: 87 bajtów
Wielkość pakietu: 545 bajtów
Wielkość pakietu: 545 bajtów
Wielkość pakietu: 209 bajtów
Wielkość pakietu: 95 bajtów
Wielkość pakietu: 91 bajtów
Wielkość pakietu: 54 bajtów
Wielkość pakietu: 94 bajtów
Wielkość pakietu: 95 bajtów
Wielkość pakietu: 98 bajtów
Wielkość pakietu: 95 bajtów
Wielkość pakietu: 98 bajtów
Wielkość pakietu: 95 bajtów
Wielkość pakietu: 98 bajtów
Wielkość pakietu: 95 bajtów
Wi

In [ ]:
# 1. pozbierac wszystko co jest do kupy [x]
# 2. zrobić "składanie i proste testy technik" [x]
# 3. wyodrębnić to na oddzielne funkcje dla pełnego porzadku [] 3 - przy okazji funkcja juz pelnego skladania na podstawie truncated []

# porównać jak tworzone flow mają się do cicflowmeter flows [X]
# zrobic rozne use case dla odczytu flow, np. na ten moment NFS i standardowo dla TCP [x] 1
# przygotowac funkcje obliczania parametrów flow [x] 2
# zająć się już stricte algorytmem [] 


# podzielić każde flow na oddzielną tablicę - dla dalszej logiki perturbacji

In [57]:
from scapy.all import IP, TCP, fragment, Padding, Raw

def change_packet_timestamp(packet, new_timestamp):
    packet.time = new_timestamp
    return packet

def add_padding_to_packet(packet, padding_size):
    if padding_size > 0:
        # Dodanie paddingu bezpośrednio do warstwy danych
        if Raw in packet:
            packet[Raw].load += b"\x00" * padding_size
        else:
            packet = packet / Raw(load=b"\x00" * padding_size)
    return packet

def fragment_packet(packet, fragment_size):
    return fragment(packet, fragsize=fragment_size)




In [188]:
from scapy.all import PcapReader, PcapWriter, IP
import os

def modify_and_write_packets_one_by_one(original_pcap, modified_pcap):
    
    if os.path.exists(modified_pcap):
        os.remove(modified_pcap)
    

    writer = PcapWriter(modified_pcap, append=True, sync=True)
    
    with PcapReader(original_pcap) as reader:
        for packet in reader:
            #TODO: dodac enumarate i sprawdzac czy w truncated packets cos takiego bedzie
            if IP in packet:
                
                #change timestamp
                #new_timestamp = packet.time + 100000000
                #packet = change_packet_timestamp(packet, new_timestamp)
                
                # fragmentation - dziala ale to dalszego sprawdzenia jak taka fragmentacja jest interpretowana TODO
                fragment_size = 500 
                packet = fragment_packet(packet, fragment_size)
                
                #padding - TODO: nie działa, do ogarnięcia
                padding_size = 200
                #packet = add_padding_before_raw(packet, padding_size)
                #packet = add_custom_protocol_with_padding(packet, padding_size)
                #print(packet)

            writer.write(packet)
            
    writer.close()


modified_pcap_path = "../data/interim/testing_small_perturbed.pcapng"
pcap_file_path = '../data/raw/testing_small.pcapng'
modify_and_write_packets_one_by_one(pcap_file_path, modified_pcap_path)


In [182]:
from scapy.all import IP, TCP, Raw
import os

def add_padding_before_raw(packet, padding_size):
    # Generuje losowe bajty jako padding
    padding = os.urandom(padding_size)
    
    if Raw in packet:
        # Pakiet ma warstwę Raw, dodaj padding przed istniejące dane
        packet[Raw].load = padding + packet[Raw].load
    else:
        # Pakiet nie ma warstwy Raw, dodaj nową warstwę Raw z paddingiem
        packet = packet / Raw(load=padding)
    
    return packet

In [185]:
from scapy.all import Ether, IP, TCP, Packet
from scapy.fields import StrField
import os

class CustomProtocol(Packet):
    name = "CustomProtocol"
    fields_desc = [StrField("padding", default="")]

def add_custom_protocol_with_padding(packet, padding_size):
    # Generuje losowe bajty jako padding
    padding = os.urandom(padding_size)
    
    # Tworzy instancję niestandardowego protokołu z paddingiem jako jego treść
    custom_protocol_instance = CustomProtocol(padding=padding)
    
    packet = packet / custom_protocol_instance
    
    return packet

In [172]:
# padding adding testing
pcap_file_path = '../data/raw/testing_small.pcapng'
with PcapReader(pcap_file_path) as reader:
    for packet in reader:
        
        print("Przed dodaniem paddingu:", len(packet))
        packet = add_custom_protocol_with_padding(packet, 200)
        print("Po dodaniu paddingu:", len(packet))
        print(packet)

        



Przed dodaniem paddingu: 66
Po dodaniu paddingu: 266
Ether / IP / TCP 192.168.56.1:59281 > 192.168.56.101:ms_wbt_server S / CustomProtocol
Przed dodaniem paddingu: 66
Po dodaniu paddingu: 266
Ether / IP / TCP 192.168.56.101:ms_wbt_server > 192.168.56.1:59281 SA / CustomProtocol
Przed dodaniem paddingu: 60
Po dodaniu paddingu: 260
Ether / IP / TCP 192.168.56.1:59281 > 192.168.56.101:ms_wbt_server A / Padding / CustomProtocol
Przed dodaniem paddingu: 98
Po dodaniu paddingu: 298
Ether / IP / TCP 192.168.56.1:59281 > 192.168.56.101:ms_wbt_server PA / Raw / CustomProtocol
Przed dodaniem paddingu: 54
Po dodaniu paddingu: 254
Ether / IP / TCP 192.168.56.101:ms_wbt_server > 192.168.56.1:59281 A / CustomProtocol
Przed dodaniem paddingu: 73
Po dodaniu paddingu: 273
Ether / IP / TCP 192.168.56.101:ms_wbt_server > 192.168.56.1:59281 PA / Raw / CustomProtocol
Przed dodaniem paddingu: 60
Po dodaniu paddingu: 260
Ether / IP / TCP 192.168.56.1:59281 > 192.168.56.101:ms_wbt_server A / Padding / CustomP

In [155]:
packet

<Ether  dst=08:00:27:e6:e5:59 src=0a:00:27:00:00:09 type=IPv4 |<IP  version=4 ihl=5 tos=0x0 len=40 id=5867 flags=DF frag=0 ttl=128 proto=tcp chksum=0xf22d src=192.168.56.1 dst=192.168.56.101 |<TCP  sport=59285 dport=ms_wbt_server seq=3310673913 ack=2356591198 dataofs=5 reserved=0 flags=RA window=0 chksum=0xdd22 urgptr=0 |<Padding  load='?\\x84\r*\\x8bP$jFwP\\xc1^[\\xf3k\\xfeO\x04\x0f\\xa6\\xf0K\x1e\\xc2\\xd6=V\\xfe\\x93\x17T\x12\x1d*\x03.I\\xc1\\xde\\xe5\\xed.\\xa5\\xa5?\\xf6\\xa1`\\xdcY\\x9e#\\xf5ҭ\x1b"z\'nӺ\\xe9\\x8f^~\\xbc\x01\\x8e\x00p\\xd6CJeR\x12\\xf3\\xe4A\\xa8`\\xd7z\\xc8TK\\xd8\x19B\nd\\x9f\x7fj(\\xa89lCR\\xc2\x1e\\x83뱎\\x9b\\xe6=\x0c\\xa85\x19\\x83\x153\\x97x\x15.Y~\\x9e\\xb3\\xa1q~\x1a`\\x96M|KlxN\\xa5\\x9a\\xc50?\\xa0\\xb17\\x89\\xad\x14\\xdcA\\x89\\xb6\x19n\x1b\x14C\\xfa\\xc2š\x15b\\xfa\x1a\\x82\\x8d\\x90\x12㱆\\xf7\x19nH\\xf9\x1a\\xa7>\\xbe\\xeb\\x84\\xe1@\\2=\\xec\\xb3{\\xb8N\x1d\\x86\\xe3\\xc1\\xe3\\xea\x00\x00\x00\x00\x00\x00' |<CustomProtocol  padding="\\xa9\\xe8ѱ\\xea

## CLEAN

In [1]:
import nfstream
import scapy

def generate_pseudo_hash(entity):
    if isinstance(entity, nfstream.flow.NFlow):
       elements = [hash(entity.src_ip), hash(entity.dst_ip), int(entity.src_port), int(entity.dst_port), int(entity.protocol)] 
    
    #elif isinstance(entity, pyshark.packet.packet.Packet): 
    #    elements = [hash(entity.ip.src), hash(entity.ip.dst), int(entity[entity.transport_layer].srcport), int(entity[entity.transport_layer].dstport), int(entity.ip.proto)]

    elif isinstance(entity, scapy.layers.l2.Ether):
        if entity.haslayer(TCP):
            elements = [hash(entity[IP].src), hash(entity[IP].dst), int(entity[TCP].sport), int(entity[TCP].dport), int(entity[IP].proto)] 
        elif entity.haslayer(UDP):
            elements = [hash(entity[IP].src), hash(entity[IP].dst), int(entity[UDP].sport), int(entity[UDP].dport),  int(entity[IP].proto)] 
    else:
        print("ERROR: not a flow nor packet")

    return sum(elements)

In [22]:
# truncated packet class

class TruncatedPacket:
    def __init__(self, packet_id, timestamp, size, pseudo_hash, flow_id, direction, src_ip, fin, tcp, udp):
        self.packet_id = packet_id
        self.timestamp = timestamp
        self.size = size
        self.pseudo_hash = pseudo_hash
        self.flow_id = flow_id
        self.direction = direction
        self.src_ip = src_ip,
        self.fin = fin,
        self.tcp = tcp,
        self.udp = udp

    def __repr__(self):
        return (f"TruncatedPacket(packet_id={self.packet_id}, "
                f"timestamp={self.timestamp}, "
                f"size={self.size}, "
                f"pseudo_hash='{self.pseudo_hash}', "
                f"flow_id={self.flow_id}, "
                f"src_ip={self.src_ip}, "
                f"fin={self.fin}, "
                f"udp={self.udp}, "
                f"tcp={self.tcp}, "
                f"direction='{self.direction}')")

In [93]:
# SCAPY TRUNCATED PACKETS GENERATION
from scapy.all import rdpcap, PcapReader, IP, TCP, UDP

def create_truncated_packets_from_pcap(file_path):
    truncated_packets = []
    cap = rdpcap(file_path)
    
    for packet_number, scapy_packet in enumerate(cap, start=1):
        if (scapy_packet.haslayer(TCP) or scapy_packet.haslayer(UDP)) and IP in scapy_packet:
            pseudo_hash = generate_pseudo_hash(scapy_packet)
            if pseudo_hash:  # we make sure that this was calculated successfully
                
                tcp = int(scapy_packet.haslayer(TCP))
                print(scapy_packet.haslayer(TCP))
                fin = False
                if tcp:
                    fin = scapy_packet[TCP].flags.FIN
                
                truncated_packet = TruncatedPacket(
                    packet_id=packet_number * 10, # according to schema
                    timestamp=scapy_packet.time, # timestamp will need to have margin of error
                    size=len(scapy_packet),
                    pseudo_hash=pseudo_hash,
                    flow_id=None,  # later
                    direction=0,  # 0 - not yet analyzed, 1 - fwd, 2 - bwd
                    src_ip=scapy_packet[IP].src,
                    # in order to assign flow id as in tcp specification, we need to record fin flags:
                    tcp=tcp,
                    udp=int(scapy_packet.haslayer(UDP)),
                    fin=fin
                )
                truncated_packets.append(truncated_packet)
    
    return truncated_packets

In [141]:
# NEW VERSION

class TruncatedPacket:
    def __init__(self, packet_id, timestamp, size, pseudo_hash, flow_id, direction, src_ip, tcp, udp, fin, syn, rst, ack, psh, urg):
        self.packet_id = packet_id
        self.timestamp = timestamp
        self.size = size
        self.pseudo_hash = pseudo_hash
        self.flow_id = flow_id
        self.direction = direction
        self.src_ip = src_ip
        self.tcp = tcp
        self.udp = udp
        self.fin = fin
        self.syn = syn
        self.rst = rst
        self.ack = ack
        self.psh = psh
        self.urg = urg

    def __repr__(self):
        return (f"TruncatedPacket(packet_id={self.packet_id}, timestamp={self.timestamp}, size={self.size}, "
                f"pseudo_hash='{self.pseudo_hash}', flow_id={self.flow_id}, direction='{self.direction}', src_ip={self.src_ip}, "
                f"tcp={self.tcp}, udp={self.udp}, fin={self.fin}, syn={self.syn}, rst={self.rst}, ack={self.ack}, psh={self.psh}, urg={self.urg})")


from scapy.all import rdpcap, IP, TCP, UDP

def create_truncated_packets_from_pcap(file_path):
    truncated_packets = []
    cap = rdpcap(file_path)
    
    for packet_number, scapy_packet in enumerate(cap, start=1):
        if IP in scapy_packet and (scapy_packet.haslayer(TCP) or scapy_packet.haslayer(UDP)):
            pseudo_hash = generate_pseudo_hash(scapy_packet)
            
            tcp, udp, fin, syn, rst, ack, psh, urg = (0,) * 8  # Initialize all flags to 0
            
            if scapy_packet.haslayer(TCP):
                tcp = 1
                flags = scapy_packet[TCP].flags
                fin = int(bool(flags & 0x01))
                syn = int(bool(flags & 0x02))
                rst = int(bool(flags & 0x04))
                ack = int(bool(flags & 0x10))
                psh = int(bool(flags & 0x08))
                urg = int(bool(flags & 0x20))
                
            udp = int(scapy_packet.haslayer(UDP))
            
            truncated_packet = TruncatedPacket(
                packet_id=packet_number * 10,  # According to schema
                timestamp=scapy_packet.time,  # Timestamp will need to have margin of error
                size=len(scapy_packet),
                pseudo_hash=pseudo_hash,
                flow_id=None,  # Later assignment
                direction=0,  # 0 - not yet analyzed, 1 - fwd, 2 - bwd
                src_ip=scapy_packet[IP].src,
                tcp=tcp,
                udp=udp,
                fin=fin,
                syn=syn,
                rst=rst,
                ack=ack,
                psh=psh,
                urg=urg
            )
            truncated_packets.append(truncated_packet)
    
    return truncated_packets

In [143]:
truncated_packets = create_truncated_packets_from_pcap('../data/raw/testing_small.pcapng')
for packet in truncated_packets:
    if packet.fin == True:
        print(packet)

In [144]:
truncated_packets

[TruncatedPacket(packet_id=10, timestamp=1595284457.571949, size=66, pseudo_hash='-3454645421489727042', flow_id=None, direction='0', src_ip=192.168.56.1, tcp=1, udp=0, fin=0, syn=1, rst=0, ack=0, psh=0, urg=0),
 TruncatedPacket(packet_id=20, timestamp=1595284457.57213, size=66, pseudo_hash='-3454645421489727042', flow_id=None, direction='0', src_ip=192.168.56.101, tcp=1, udp=0, fin=0, syn=1, rst=0, ack=1, psh=0, urg=0),
 TruncatedPacket(packet_id=30, timestamp=1595284457.572317, size=60, pseudo_hash='-3454645421489727042', flow_id=None, direction='0', src_ip=192.168.56.1, tcp=1, udp=0, fin=0, syn=0, rst=0, ack=1, psh=0, urg=0),
 TruncatedPacket(packet_id=40, timestamp=1595284457.573703, size=98, pseudo_hash='-3454645421489727042', flow_id=None, direction='0', src_ip=192.168.56.1, tcp=1, udp=0, fin=0, syn=0, rst=0, ack=1, psh=1, urg=0),
 TruncatedPacket(packet_id=50, timestamp=1595284457.62409, size=54, pseudo_hash='-3454645421489727042', flow_id=None, direction='0', src_ip=192.168.56.

In [5]:
# START SPOSÓB NA PRZYPISYWANIE PAKIETÓW DO FLOW - ZGODNIE Z NFSTREAM SPEC

def assign_flow_ids_to_packets(truncated_packets, streamer):
    for flow in streamer:
        for packet in truncated_packets:
            packet_timestamp = float(packet.timestamp) * 1000
            flow_firstseen = float(flow.bidirectional_first_seen_ms)
            flow_lastseen = float(flow.bidirectional_last_seen_ms)
            
            
            if flow_firstseen - 1.0 <= packet_timestamp <= flow_lastseen + 1.0 and packet.pseudo_hash == generate_pseudo_hash(flow):
                packet.flow_id = flow.id
                if packet.src_ip == flow.src_ip:
                    packet.direction = 1
                else:
                    packet.direction = 2

In [147]:
TCP_EXPIRATION = 240.0
UDP_EXPIRATION = 240.0


def assign_flow_ids_to_packets(truncated_packets):
    from collections import defaultdict

    # Group packets by pseudo_hash
    packets_by_hash = defaultdict(list)
    for packet in truncated_packets:
        packets_by_hash[packet.pseudo_hash].append(packet)

    global_flow_id = 1  # Initialize global flow ID

    for hash_group in packets_by_hash.values():
        # Sort packets within each group by timestamp
        hash_group.sort(key=lambda pkt: pkt.timestamp)

        first_packet_src_ip = None  # Source IP of the first packet in a flow
        last_timestamp = None  # Timestamp of the last packet
        waiting_for_new_flow = False  # Indicates if we are waiting to start a new flow

        for i, packet in enumerate(hash_group):
            # Check conditions for ending the current flow
            timeout_condition = (packet.tcp and last_timestamp is not None and (packet.timestamp - last_timestamp) > TCP_EXPIRATION) or \
                                (packet.udp and last_timestamp is not None and (packet.timestamp - last_timestamp) > UDP_EXPIRATION)
            fin_condition = packet.tcp and packet.fin and not waiting_for_new_flow

            if timeout_condition or fin_condition:
                if timeout_condition:  # New flow starts after a timeout, not after FIN
                    waiting_for_new_flow = True
                if fin_condition:  # After FIN, we're still in the same flow
                    waiting_for_new_flow = False

            if waiting_for_new_flow and not packet.fin:  # New flow starts after FIN or timeout
                global_flow_id += 1
                waiting_for_new_flow = False  # Reset waiting for new flow

            if not first_packet_src_ip or waiting_for_new_flow:  # Setting for a new flow
                first_packet_src_ip = packet.src_ip

            packet.flow_id = global_flow_id
            packet.direction = 1 if packet.src_ip == first_packet_src_ip else 2

            last_timestamp = packet.timestamp  # Update the timestamp of the last packet

            # If it's the last packet in the hash group, ensure we end the current flow
            if i == len(hash_group) - 1:
                global_flow_id += 1
                waiting_for_new_flow = False

    return truncated_packets


In [148]:
assign_flow_ids_to_packets(truncated_packets)

[TruncatedPacket(packet_id=10, timestamp=1595284457.571949, size=66, pseudo_hash='-3454645421489727042', flow_id=1, direction='1', src_ip=192.168.56.1, tcp=1, udp=0, fin=0, syn=1, rst=0, ack=0, psh=0, urg=0),
 TruncatedPacket(packet_id=20, timestamp=1595284457.57213, size=66, pseudo_hash='-3454645421489727042', flow_id=1, direction='2', src_ip=192.168.56.101, tcp=1, udp=0, fin=0, syn=1, rst=0, ack=1, psh=0, urg=0),
 TruncatedPacket(packet_id=30, timestamp=1595284457.572317, size=60, pseudo_hash='-3454645421489727042', flow_id=1, direction='1', src_ip=192.168.56.1, tcp=1, udp=0, fin=0, syn=0, rst=0, ack=1, psh=0, urg=0),
 TruncatedPacket(packet_id=40, timestamp=1595284457.573703, size=98, pseudo_hash='-3454645421489727042', flow_id=1, direction='1', src_ip=192.168.56.1, tcp=1, udp=0, fin=0, syn=0, rst=0, ack=1, psh=1, urg=0),
 TruncatedPacket(packet_id=50, timestamp=1595284457.62409, size=54, pseudo_hash='-3454645421489727042', flow_id=1, direction='2', src_ip=192.168.56.101, tcp=1, udp

In [125]:
def summarize_flow_statistics(truncated_packets):
    from collections import defaultdict

    # Initialize a dictionary to hold statistics for each flow
    flow_statistics = defaultdict(lambda: {
        'packet_count': 0, 
        'total_size': 0, 
        'start_time': None, 
        'end_time': None, 
        'src_ips': set()  # Use a set to avoid duplicate IPs
    })

    for packet in truncated_packets:
        flow_id = packet.flow_id
        flow_stats = flow_statistics[flow_id]
        
        flow_stats['packet_count'] += 1
        flow_stats['total_size'] += packet.size
        flow_stats['src_ips'].add(packet.src_ip[0])  # Add src_ip to the set
        
        # Update start and end time for the flow
        if flow_stats['start_time'] is None or packet.timestamp < flow_stats['start_time']:
            flow_stats['start_time'] = packet.timestamp
        if flow_stats['end_time'] is None or packet.timestamp > flow_stats['end_time']:
            flow_stats['end_time'] = packet.timestamp

    # Print summarized statistics for each flow
    for flow_id, stats in flow_statistics.items():
        duration = stats['end_time'] - stats['start_time'] if stats['start_time'] and stats['end_time'] else 0
        src_ips = ", ".join(stats['src_ips'])  # Convert set of IPs to string for printing
        print(f"Flow ID: {flow_id}, Packet Count: {stats['packet_count']}, Total Size: {stats['total_size']}, Duration: {duration}, Src IPs: {src_ips}")

summarize_flow_statistics(truncated_packets)


Flow ID: 1, Packet Count: 278, Total Size: 29798, Duration: 14.052047, Src IPs: 192.168.56.1, 192.168.56.101
Flow ID: 2, Packet Count: 2, Total Size: 940, Duration: 0.009462, Src IPs: 192.168.56.101, 192.168.56.100
Flow ID: 3, Packet Count: 4, Total Size: 384, Duration: 0.031884, Src IPs: 192.168.56.101
Flow ID: 4, Packet Count: 1, Total Size: 71, Duration: 0.00000, Src IPs: 192.168.56.101
Flow ID: 5, Packet Count: 1, Total Size: 71, Duration: 0.000000, Src IPs: 192.168.56.101
Flow ID: 6, Packet Count: 6, Total Size: 1074, Duration: 15.170359, Src IPs: 192.168.56.101
Flow ID: 7, Packet Count: 217, Total Size: 72273, Duration: 11.597789, Src IPs: 192.168.56.1, 192.168.56.101
Flow ID: 8, Packet Count: 289, Total Size: 30707, Duration: 11.549842, Src IPs: 192.168.56.1, 192.168.56.101
Flow ID: 9, Packet Count: 230, Total Size: 95484, Duration: 9.220849, Src IPs: 192.168.56.1, 192.168.56.101


In [127]:
import numpy as np

def calculate_size_statistics(flow_id, truncated_packets, direction=0):
    sizes = [pkt.size for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction)]
    
    if not sizes:
        return "No packets found matching criteria."
    
    size_mean = np.mean(sizes)
    size_min = np.min(sizes)
    size_max = np.max(sizes)
    size_std = np.std(sizes)
    
    return {
        'mean': size_mean,
        'min': size_min,
        'max': size_max,
        'std': size_std
    }


In [149]:
calculate_size_statistics(1, truncated_packets, direction=1)

{'mean': 107.59638554216868, 'min': 60, 'max': 762, 'std': 91.83005380656388}

In [138]:
def calculate_delta_time_statistics(flow_id, truncated_packets, direction=0):
    import numpy as np

    timestamps = [pkt.timestamp for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction)]
    
    if len(timestamps) < 2:
        return "Insufficient packets for delta time calculation."
    
    # Sorting timestamps before calculating deltas to ensure chronological order
    timestamps = sorted(timestamps)
    # Calculating time deltas between consecutive packets
    delta_times = np.diff(timestamps)

    # Convert delta_times to a numpy array of type float to ensure compatibility with numpy operations
    delta_times = np.array(delta_times, dtype=float)
    
    delta_time_mean = np.mean(delta_times)
    delta_time_min = np.min(delta_times)
    delta_time_max = np.max(delta_times)
    delta_time_std = np.std(delta_times)
    
    return {
        'mean': delta_time_mean,
        'min': delta_time_min,
        'max': delta_time_max,
        'std': delta_time_std
    }


In [140]:
calculate_delta_time_statistics(1, truncated_packets, direction=0)

{'mean': 0.050729411552346566,
 'min': 2e-06,
 'max': 1.638539,
 'std': 0.16143004716714476}

In [150]:
def total_flow_size(flow_id, truncated_packets, direction=0):
    total_size = sum(pkt.size for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction))
    return total_size


In [153]:
total_flow_size(1, truncated_packets, direction=0)

29798

In [152]:
def total_flow_duration(flow_id, truncated_packets, direction=0):
    timestamps = [pkt.timestamp for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction)]
    if timestamps:
        return max(timestamps) - min(timestamps)
    else:
        return "No packets found matching criteria."


In [154]:
total_flow_duration(1, truncated_packets, direction=0)

Decimal('14.052047')

In [155]:
def total_packet_count(flow_id, truncated_packets, direction=0):
    count = sum(1 for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction))
    return count


In [156]:
total_packet_count(1, truncated_packets)


278

In [160]:
def count_tcp_flags(flow_id, truncated_packets, direction=0):
    flags_count = {'FIN': 0, 'SYN': 0, 'RST': 0, 'PSH': 0, 'ACK': 0, 'URG': 0}
    for pkt in truncated_packets:
        if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction) and pkt.tcp:
            flags_count['FIN'] += pkt.fin
            flags_count['SYN'] += pkt.syn
            flags_count['RST'] += pkt.rst
            flags_count['PSH'] += pkt.psh
            flags_count['ACK'] += pkt.ack
            flags_count['URG'] += pkt.urg
    return flags_count


In [161]:
count_tcp_flags(1, truncated_packets)

{'FIN': 0, 'SYN': 2, 'RST': 1, 'PSH': 189, 'ACK': 277, 'URG': 0}

In [33]:
from nfstream import NFStreamer

pcap_file_path = '../data/raw/testing_small.pcapng'

truncated_packets = create_truncated_packets_from_pcap(pcap_file_path)

streamer = NFStreamer(source=pcap_file_path)
assign_flow_ids_to_packets(truncated_packets, streamer)



TypeError: assign_flow_ids_to_packets() takes 1 positional argument but 2 were given

In [17]:
# TEST

for packet in truncated_packets:
    print(packet.timestamp)

TruncatedPacket(packet_id=10, timestamp=1595284457.571949, size=66, pseudo_hash='-331772846983708113', flow_id=8, src_ip=192.168.56.1, direction='1')
TruncatedPacket(packet_id=20, timestamp=1595284457.57213, size=66, pseudo_hash='-331772846983708113', flow_id=8, src_ip=192.168.56.101, direction='2')
TruncatedPacket(packet_id=30, timestamp=1595284457.572317, size=60, pseudo_hash='-331772846983708113', flow_id=8, src_ip=192.168.56.1, direction='1')
TruncatedPacket(packet_id=40, timestamp=1595284457.573703, size=98, pseudo_hash='-331772846983708113', flow_id=8, src_ip=192.168.56.1, direction='1')
TruncatedPacket(packet_id=50, timestamp=1595284457.62409, size=54, pseudo_hash='-331772846983708113', flow_id=8, src_ip=192.168.56.101, direction='2')
TruncatedPacket(packet_id=60, timestamp=1595284457.648739, size=73, pseudo_hash='-331772846983708113', flow_id=8, src_ip=192.168.56.101, direction='2')
TruncatedPacket(packet_id=70, timestamp=1595284457.689521, size=60, pseudo_hash='-33177284698370

# super clean

In [ ]:
# truncated packet class

#from .flow import generate_pseudo_hash

class TruncatedPacket:
    def __init__(self, packet_id, timestamp, size, pseudo_hash, flow_id, direction, src_ip, tcp, udp, fin, syn, rst, ack, psh, urg):
        self.packet_id = packet_id
        self.timestamp = timestamp
        self.size = size
        self.pseudo_hash = pseudo_hash
        self.flow_id = flow_id
        self.direction = direction
        self.src_ip = src_ip
        self.tcp = tcp
        self.udp = udp
        self.fin = fin
        self.syn = syn
        self.rst = rst
        self.ack = ack
        self.psh = psh
        self.urg = urg

    def __repr__(self):
        return (f"TruncatedPacket(packet_id={self.packet_id}, timestamp={self.timestamp}, size={self.size}, "
                f"pseudo_hash='{self.pseudo_hash}', flow_id={self.flow_id}, direction='{self.direction}', src_ip={self.src_ip}, "
                f"tcp={self.tcp}, udp={self.udp}, fin={self.fin}, syn={self.syn}, rst={self.rst}, ack={self.ack}, psh={self.psh}, urg={self.urg})")


from scapy.all import rdpcap, IP, TCP, UDP

def create_truncated_packets_from_pcap(file_path):
    truncated_packets = []
    cap = rdpcap(file_path)
    
    for packet_number, scapy_packet in enumerate(cap, start=1):
        if IP in scapy_packet and (scapy_packet.haslayer(TCP) or scapy_packet.haslayer(UDP)):
            pseudo_hash = generate_pseudo_hash(scapy_packet)
            
            tcp, udp, fin, syn, rst, ack, psh, urg = (0,) * 8  # Initialize all flags to 0
            
            if scapy_packet.haslayer(TCP):
                tcp = 1
                flags = scapy_packet[TCP].flags
                fin = int(bool(flags & 0x01))
                syn = int(bool(flags & 0x02))
                rst = int(bool(flags & 0x04))
                ack = int(bool(flags & 0x10))
                psh = int(bool(flags & 0x08))
                urg = int(bool(flags & 0x20))
                
            udp = int(scapy_packet.haslayer(UDP))
            
            truncated_packet = TruncatedPacket(
                packet_id=packet_number * 10,  # According to schema
                timestamp=scapy_packet.time,  # Timestamp will need to have margin of error
                size=len(scapy_packet),
                pseudo_hash=pseudo_hash,
                flow_id=None,  # Later assignment
                direction=0,  # 0 - not yet analyzed, 1 - fwd, 2 - bwd
                src_ip=scapy_packet[IP].src,
                tcp=tcp,
                udp=udp,
                fin=fin,
                syn=syn,
                rst=rst,
                ack=ack,
                psh=psh,
                urg=urg
            )
            truncated_packets.append(truncated_packet)
    
    return truncated_packets

In [ ]:
# flow_utilities

import nfstream
import scapy

TCP_EXPIRATION = 240.0
UDP_EXPIRATION = 240.0

def generate_pseudo_hash(entity):
    if isinstance(entity, nfstream.flow.NFlow):
       elements = [hash(entity.src_ip), hash(entity.dst_ip), int(entity.src_port), int(entity.dst_port), int(entity.protocol)] 
    
    #elif isinstance(entity, pyshark.packet.packet.Packet): 
    #    elements = [hash(entity.ip.src), hash(entity.ip.dst), int(entity[entity.transport_layer].srcport), int(entity[entity.transport_layer].dstport), int(entity.ip.proto)]

    elif isinstance(entity, scapy.layers.l2.Ether):
        if entity.haslayer(TCP):
            elements = [hash(entity[IP].src), hash(entity[IP].dst), int(entity[TCP].sport), int(entity[TCP].dport), int(entity[IP].proto)] 
        elif entity.haslayer(UDP):
            elements = [hash(entity[IP].src), hash(entity[IP].dst), int(entity[UDP].sport), int(entity[UDP].dport),  int(entity[IP].proto)] 
    else:
        print("ERROR: not a flow nor packet")

    return sum(elements)

def assign_flow_ids_to_packets(truncated_packets):
    from collections import defaultdict

    # Group packets by pseudo_hash
    packets_by_hash = defaultdict(list)
    for packet in truncated_packets:
        packets_by_hash[packet.pseudo_hash].append(packet)

    global_flow_id = 1  # Initialize global flow ID

    for hash_group in packets_by_hash.values():
        # Sort packets within each group by timestamp
        hash_group.sort(key=lambda pkt: pkt.timestamp)

        first_packet_src_ip = None  # Source IP of the first packet in a flow
        last_timestamp = None  # Timestamp of the last packet
        waiting_for_new_flow = False  # Indicates if we are waiting to start a new flow

        for i, packet in enumerate(hash_group):
            # Check conditions for ending the current flow
            timeout_condition = (packet.tcp and last_timestamp is not None and (packet.timestamp - last_timestamp) > TCP_EXPIRATION) or \
                                (packet.udp and last_timestamp is not None and (packet.timestamp - last_timestamp) > UDP_EXPIRATION)
            fin_condition = packet.tcp and packet.fin and not waiting_for_new_flow

            if timeout_condition or fin_condition:
                if timeout_condition:  # New flow starts after a timeout, not after FIN
                    waiting_for_new_flow = True
                if fin_condition:  # After FIN, we're still in the same flow
                    waiting_for_new_flow = False

            if waiting_for_new_flow and not packet.fin:  # New flow starts after FIN or timeout
                global_flow_id += 1
                waiting_for_new_flow = False  # Reset waiting for new flow

            if not first_packet_src_ip or waiting_for_new_flow:  # Setting for a new flow
                first_packet_src_ip = packet.src_ip

            packet.flow_id = global_flow_id
            packet.direction = 1 if packet.src_ip == first_packet_src_ip else 2

            last_timestamp = packet.timestamp  # Update the timestamp of the last packet

            # If it's the last packet in the hash group, ensure we end the current flow
            if i == len(hash_group) - 1:
                global_flow_id += 1
                waiting_for_new_flow = False

    return truncated_packets


In [ ]:
# flow calculations

import numpy as np

def calculate_size_statistics(flow_id, truncated_packets, direction=0):
    sizes = [pkt.size for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction)]
    
    if not sizes:
        return "No packets found matching criteria."
    
    size_mean = np.mean(sizes)
    size_min = np.min(sizes)
    size_max = np.max(sizes)
    size_std = np.std(sizes)
    
    return {
        'mean': size_mean,
        'min': size_min,
        'max': size_max,
        'std': size_std
    }

def calculate_delta_time_statistics(flow_id, truncated_packets, direction=0):
    import numpy as np

    timestamps = [pkt.timestamp for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction)]
    
    if len(timestamps) < 2:
        return "Insufficient packets for delta time calculation."
    
    # Sorting timestamps before calculating deltas to ensure chronological order
    timestamps = sorted(timestamps)
    # Calculating time deltas between consecutive packets
    delta_times = np.diff(timestamps)

    # Convert delta_times to a numpy array of type float to ensure compatibility with numpy operations
    delta_times = np.array(delta_times, dtype=float)
    
    delta_time_mean = np.mean(delta_times)
    delta_time_min = np.min(delta_times)
    delta_time_max = np.max(delta_times)
    delta_time_std = np.std(delta_times)
    
    return {
        'mean': delta_time_mean,
        'min': delta_time_min,
        'max': delta_time_max,
        'std': delta_time_std
    }
    
    
def total_flow_size(flow_id, truncated_packets, direction=0):
    total_size = sum(pkt.size for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction))
    return total_size


def total_flow_duration(flow_id, truncated_packets, direction=0):
    timestamps = [pkt.timestamp for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction)]
    if timestamps:
        return max(timestamps) - min(timestamps)
    else:
        return "No packets found matching criteria."

def total_packet_count(flow_id, truncated_packets, direction=0):
    count = sum(1 for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction))
    return count

def total_packet_count(flow_id, truncated_packets, direction=0):
    count = sum(1 for pkt in truncated_packets if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction))
    return count


def count_tcp_flags(flow_id, truncated_packets, direction=0):
    flags_count = {'FIN': 0, 'SYN': 0, 'RST': 0, 'PSH': 0, 'ACK': 0, 'URG': 0}
    for pkt in truncated_packets:
        if pkt.flow_id == flow_id and (direction == 0 or pkt.direction == direction) and pkt.tcp:
            flags_count['FIN'] += pkt.fin
            flags_count['SYN'] += pkt.syn
            flags_count['RST'] += pkt.rst
            flags_count['PSH'] += pkt.psh
            flags_count['ACK'] += pkt.ack
            flags_count['URG'] += pkt.urg
    return flags_count




In [226]:
# restoration utilities

import os
from scapy.all import Ether, IP, TCP, Packet, fragment, Padding, Raw, PcapReader, PcapWriter
from scapy.fields import StrField
    
class StealthProtocol(Packet):
    name = "StealthProtocol"
    fields_desc = [StrField("padding", default="")]

def add_padding(packet, padding_size):
    padding = os.urandom(padding_size)
    
    stealth_protocol_instance = StealthProtocol(padding=padding)
    
    packet = packet / stealth_protocol_instance
    
    return packet


def change_packet_timestamp(packet, new_timestamp):
    packet.time = new_timestamp
    return packet


def fragment_packet(packet, fragment_size):
    return fragment(packet, fragsize=fragment_size)


def split_flow(packet):
    # Ensure the packet is an IP packet containing a TCP layer
    if IP in packet and TCP in packet:
        # Create a new packet with the FIN flag, copying IP addresses and ports from the original packet
        fin_packet = Ether(src=packet[Ether].src, dst=packet[Ether].dst)/IP(src=packet[IP].src, dst=packet[IP].dst)/TCP(sport=packet[TCP].sport, dport=packet[TCP].dport, flags="F")
        fin_packet.time = packet.time + 0.00001
        print(f'fin:{fin_packet.time}')
        return fin_packet
    # Else it has to be UDP:)
    else:
        fin_packet = Ether(src=packet[Ether].src, dst=packet[Ether].dst)/IP(src=packet[IP].src, dst=packet[IP].dst)/TCP(sport=packet[UDP].sport, dport=packet[UDP].dport, flags="F")
        return fin_packet

def dummy_packet(packet, size):
    if IP in packet:
        dummy = Ether(src=packet[Ether].src, dst=packet[Ether].dst) / IP(src=packet[IP].src, dst=packet[IP].dst)
        
        if TCP in packet:
            dummy /= TCP(sport=packet[TCP].sport, dport=packet[TCP].dport)
        elif UDP in packet:
            dummy /= UDP(sport=packet[UDP].sport, dport=packet[UDP].dport)
        else:
            return "The packet does not contain a TCP or UDP layer."
        
        # Add a payload to meet the desired size, considering the size of the headers
        payload_size = max(0, size - len(dummy))
        dummy /= ("0" * payload_size)
        dummy.time = packet.time + 0.000001
        
        return dummy
    else:
        return "The packet does not contain an IP layer."
    
#TODO: maybe block dummy_flags for UDP flows
def dummy_flag(packet, flags):
    if IP in packet:
        src_port, dst_port = 0, 0
        
        if TCP in packet:
            src_port, dst_port = packet[TCP].sport, packet[TCP].dport
        elif UDP in packet:
            src_port, dst_port = packet[UDP].sport, packet[UDP].dport
        else:
            return "The packet does not contain a TCP or UDP layer."
        
        # Generate a TCP packet with the specified flags, using ports from the original packet
        flagged_packet = Ether(src=packet[Ether].src, dst=packet[Ether].dst) / IP(src=packet[IP].src, dst=packet[IP].dst)/TCP(sport=src_port, dport=dst_port, flags=flags)
        flagged_packet.time = packet.time + 0.000001
        
        return flagged_packet
    else:
        return "The packet does not contain an IP layer."
    
    
def modify_and_write_packets_one_by_one(original_pcap, modified_pcap, truncated_packets=None):
    
    if os.path.exists(modified_pcap):
        os.remove(modified_pcap)
    

    writer = PcapWriter(modified_pcap, append=True, sync=True)
    truncated_packets_dict = {tp.packet_id: tp for tp in truncated_packets}
    
    with PcapReader(original_pcap) as reader:
        for packet in reader:
            packet_id = int(packet[IP].id)  # Assuming IP.id as a key for simplicity
            if packet_id in truncated_packets_dict:
                tp = truncated_packets_dict[packet_id]
                
                # Change timestamp if necessary
                packet.time = tp.timestamp
                
                # Adjust packet size by adding padding or fragmenting
                packet_length = len(packet)
                if packet_length < tp.size:
                    packet = add_padding(packet, tp.size - packet_length)
                elif packet_length > tp.size:
                    packet = fragment_packet(packet, tp.size)[0]  # Take the first fragment for simplicity
                
                # TODO: Implement conditions for dummy_packet, split_flow, or dummy_flag

                    #if dla od (packet_id * 10 + 5) do (packet_id * 10 + 9) istnieja jakies truncated packets o takich indeksach to:
                    #    dummy_packet albo flow_split albo dummy_flag to jeszcze to zaimplementowania 
                        
                
                #TEST CASE SCENARIOS WHICH WORKED:
                
                #CHANGE TIMESTAMP
                #new_timestamp = packet.time + 100000000
                #packet = change_packet_timestamp(packet, new_timestamp)
                
                #PACKET FRAGMENTATION
                #fragment_size = 500 
                #packet = fragment_packet(packet, fragment_size)

                #PADDING
                #padding_size = 200
                #packet = add_padding(packet, padding_size)
                
                #SPLIT FLOW
                #writer.write(packet) - dont send afterwards
                #splt = split_flow(packet)
                #writer.write(splt)
                    
                #DUMMY PACKET
                #writer.write(packet) - dont send afterwards
                #writer.write(dummy_packet(packet, 1000))
                
                #DUMMY FLAG
                #writer.write(packet) - donr send afterwards
                #writer.write(dummy_flag(packet, "SA"))

            writer.write(packet)
            
    writer.close()
    


In [228]:
# main

modified_pcap_path = "../data/interim/testing_small_perturbed.pcapng"
pcap_file_path = '../data/raw/testing_small.pcapng'

truncated_packets = create_truncated_packets_from_pcap(pcap_file_path)
truncated_packets = assign_flow_ids_to_packets(truncated_packets)
modify_and_write_packets_one_by_one(pcap_file_path, modified_pcap_path, truncated_packets)

In [1]:
import sys, os
sys.path.append(os.path.abspath('../'))

In [17]:
# saved files testing
from src.utils.flow import *
from src.utils.flow_calculations import *
from src.utils.restoration import *
from src.utils.truncated_packet import *

modified_pcap_path = "../data/interim/testing_small_perturbed.pcapng"
pcap_file_path = '../data/raw/testing_small.pcapng'

truncated_packets = create_truncated_packets_from_pcap(pcap_file_path)
truncated_packets = assign_flow_ids_to_packets(truncated_packets)
print(calculate_delta_time_statistics(7, truncated_packets, direction=0))
modify_and_write_packets_one_by_one(pcap_file_path, modified_pcap_path, truncated_packets)

{'mean': 0.053693467592592595, 'min': 1e-06, 'max': 1.95455, 'std': 0.1967639528377633}
